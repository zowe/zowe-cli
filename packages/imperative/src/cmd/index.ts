/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./doc/profiles/definition/ICommandProfileSchema";
export * from "./doc/profiles/definition/ICommandProfileProperty";
export * from "./doc/profiles/definition/ICommandProfileTypeConfiguration";
export * from "./doc/handler/ICommandHandler";
export * from "./doc/handler/ICommandHandlerResponseChecker";
export * from "./doc/handler/ICommandHandlerResponseValidator";
export * from "./doc/handler/IHandlerParameters";
export * from "./doc/handler/IChainedHandlerEntry";
export * from "./doc/handler/IChainedHandlerArgumentMapping";

export * from "./doc/option/ICommandOptionAllowableValues";
export * from "./doc/option/ICommandOptionDefinition";
export * from "./doc/option/ICommandOptionValueImplications";
export * from "./doc/option/ICommandPositionalDefinition";

export * from "./doc/response/response/ICommandResponse";
export * from "./doc/response/parms/ICommandResponseParms";
export * from "./doc/response/response/ICommandValidatorError";
export * from "./doc/response/response/ICommandValidatorResponse";

export * from "./doc/ICommandDefinition";
export * from "./doc/ICommandDefinitionPassOn";
export * from "./doc/ICommandExampleDefinition";
export * from "./doc/IPartialCommandDefinition";

export * from "./doc/args/ICommandArguments";

export * from "./handlers/FailedCommandHandler";

export * from "./help/abstract/AbstractHelpGenerator";
export * from "./help/abstract/AbstractHelpGeneratorFactory";

export * from "./help/doc/IHelpGeneratorFactory";
export * from "./help/doc/IHelpGeneratorFactoryParms";
export * from "./help/doc/IHelpGeneratorParms";
export * from "./help/DefaultHelpGenerator";
export * from "./help/HelpConstants";
export * from "./help/HelpGeneratorFactory";
export * from "./help/doc/IHelpGenerator";
export * from "./help/WebHelpGenerator";
export * from "./help/WebHelpManager";

export * from "./doc/profiles/definition/ICommandProfile";
export * from "./doc/profiles/definition/ICommandProfileTypeConfiguration";

export * from "./profiles/CliProfileManager";

export * from "./syntax/SyntaxValidator";

export * from "./utils/CommandUtils";
export * from "./utils/SharedOptions";

export * from "./yargs/doc/IYargsParms";
export * from "./yargs/doc/IYargsResponse";

export * from "./yargs/AbstractCommandYargs";
export * from "./yargs/CommandYargs";
export * from "./yargs/GroupCommandYargs";
export * from "./yargs/YargsConfigurer";
export * from "./yargs/YargsDefiner";

export * from "./CommandPreparer";
export * from "./CommandProcessor";
export * from "./response/CommandResponse";

export * from "./profiles/CommandProfiles";

export * from "./response/CommandResponse";
export * from "./response/HandlerResponse";

export * from "./doc/response/api/handler/IHandlerResponseApi";
export * from "./doc/response/api/handler/IHandlerProgressApi";
export * from "./doc/response/api/handler/IHandlerResponseConsoleApi";
export * from "./doc/response/api/handler/IHandlerResponseDataApi";
export * from "./doc/response/api/handler/IHandlerFormatOutputApi";

export * from "./doc/response/response/ICommandOutputFormat";
