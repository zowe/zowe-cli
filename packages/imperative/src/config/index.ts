/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./Config";
export * from "./ConfigAutoStore";
export * from "./ConfigConstants";
export * from "./ConfigSchema";
export * from "./ConfigBuilder";
export * as ConfigUtils from "./ConfigUtils";
export * from "./ProfileCredentials";
export * from "./ProfileInfo";
export * from "./ProfInfoErr";
export * from "./doc/IConfig";
export * from "./doc/IConfigAutoStoreOpts";
export * from "./doc/IConfigBuilderOpts";
export * from "./doc/IConfigConvertResult";
export * from "./doc/IConfigLayer";
export * from "./doc/IConfigOpts";
export * from "./doc/IConfigProfile";
export * from "./doc/IConfigSchema";
export * from "./doc/IConfigSecure";
export * from "./doc/IConfigVault";
export * from "./doc/IProfArgAttrs";
export * from "./doc/IProfAttrs";
export * from "./doc/IProfInfoUpdatePropOpts";
export * from "./doc/IProfInfoRemoveKnownPropOpts";
export * from "./doc/IProfInfoErrParms";
export * from "./doc/IProfLoc";
export * from "./doc/IProfMergeArgOpts";
export * from "./doc/IProfMergedArg";
export * from "./doc/IProfInfoProps";
export * from "./doc/IProfOpts";
