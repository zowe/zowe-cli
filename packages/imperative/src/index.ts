/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./censor";
export * from "./cmd";
export * from "./config";
export * from "./console";
export * from "./constants";
export * from "./events";
export * from "./error";
export * from "./expect";
export * from "./imperative";
// export * from "./interfaces";
export * from "./io";
export * from "./logger";
export * from "./messages";
export * from "./operations";
export * from "./profiles";
export * from "./rest";
export * from "./security";
export * from "./settings";
export * from "./utilities";
