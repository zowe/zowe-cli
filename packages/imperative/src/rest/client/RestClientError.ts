/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeError } from "../../error";
import { IRestClientError } from "./doc/IRestClientError";
import { IImperativeErrorParms } from "../../error/doc/IImperativeErrorParms";

export class RestClientError extends ImperativeError {
    /**
     * Creates an instance of RestClientError.
     * @param {IRestClientError} mDetails - The IRestClientError details provided by the rest client. Contains
     *                                      "relevant" diagnostic information such as host/port/request details, etc.
     * @param {IImperativeErrorParms} [parms] - Imperative error parameters.
     */
    constructor(public mDetails: IRestClientError, parms?: IImperativeErrorParms) {
        super(mDetails, parms);
    }

    /**
     * Accessor for IRestClientError error details.
     * @type {IRestClientError}
     */
    public get details(): IRestClientError {
        return this.mDetails;
    }
}
