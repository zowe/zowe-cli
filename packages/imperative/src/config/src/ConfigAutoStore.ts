/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { ICommandArguments, IHandlerParameters } from "../../cmd";
import { ICommandHandlerRequire } from "../../cmd/src/doc/handler/ICommandHandlerRequire";
import { ICommandProfileAuthConfig } from "../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";
import { ConfigUtils } from "./ConfigUtils";
import { AbstractAuthHandler } from "../../imperative/src/auth/handlers/AbstractAuthHandler";
import { ImperativeConfig } from "../../utilities";
import { ISession } from "../../rest/src/session/doc/ISession";
import { Session } from "../../rest/src/session/Session";
import { AUTH_TYPE_TOKEN, TOKEN_TYPE_APIML } from "../../rest/src/session/SessConstants";
import { Logger } from "../../logger";
import {
    IConfigAutoStoreFindActiveProfileOpts,
    IConfigAutoStoreFindAuthHandlerForProfileOpts,
    IConfigAutoStoreStoreSessCfgPropsOpts
} from "./doc/IConfigAutoStoreOpts";

/**
 * Class to manage automatic storage of properties in team config.
 */
export class ConfigAutoStore {
    /**
     * Finds the profile where auto-store properties should be saved.
     * @param params CLI handler parameters object
     * @param profileProps List of properties required in the profile schema
     * @returns Tuple containing profile type and name, or undefined if no profile was found
     */
    public static findActiveProfile(params: IHandlerParameters, profileProps: string[]): [string, string] | undefined {
        return this._findActiveProfile({ params, profileProps });
    }
    /**
     * Helper method to find an active profile based on the optional CLI handler parameters
     * @param opts Set of options required to find an active profile
     * @returns Tuple containing profile type and name, or undefined if no profile was found
     */
    private static _findActiveProfile(opts: IConfigAutoStoreFindActiveProfileOpts): [string, string] | undefined {
        const profileTypes = typeof opts.params !== "undefined" ? [
            ...(opts.params.definition.profile?.required || []),
            ...(opts.params.definition.profile?.optional || [])
        ] : opts.profileTypes || [];

        for (const profType of profileTypes) {
            const profileMatch = ImperativeConfig.instance.loadedConfig.profiles?.find(p => p.type === profType);
            if (profileMatch != null && opts.profileProps.every(propName => propName in profileMatch.schema.properties)) {
                return [profType, ConfigUtils.getActiveProfileName(profType, opts.params?.arguments, opts.defaultProfileName)];
            }
        }
    }

    /**
     * Finds the token auth handler class for a team config profile.
     * @param profilePath JSON path of profile
     * @param cmdArguments CLI arguments which may specify a profile
     * @returns Auth handler class or undefined if none was found
     */
    public static findAuthHandlerForProfile(profilePath: string, cmdArguments: ICommandArguments): AbstractAuthHandler | undefined {
        return this._findAuthHandlerForProfile({ profilePath, cmdArguments });
    }

    /**
     * Helper method that finds the token auth handler class for a team config profile
     * @param opts Set of options required to find the auth handler for a given profile path
     * @returns Auth handler class or undefined if none was found
     */
    private static _findAuthHandlerForProfile(opts: IConfigAutoStoreFindAuthHandlerForProfileOpts): AbstractAuthHandler | undefined {
        const config = opts.config || ImperativeConfig.instance.config;
        const profileType = lodash.get(config.properties, `${opts.profilePath}.type`);
        const profile = config.api.profiles.get(opts.profilePath.replace(/profiles\./g, ""), false);

        if (profile == null || profileType == null) { // Profile must exist and have type defined
            return;
        } else if (profileType === "base") {
            if (profile.tokenType == null) { // Base profile must have tokenType defined
                return;
            }
        } else {
            if (profile.basePath == null) { // Service profiles must have basePath defined
                return;
            }
            if (profile.tokenType == null) {  // If tokenType undefined in service profile, fall back to base profile
                const baseProfileName = ConfigUtils.getActiveProfileName("base", opts.cmdArguments, opts.defaultBaseProfileName);
                return this._findAuthHandlerForProfile({ ...opts, profilePath: config.api.profiles.getProfilePathFromName(baseProfileName) });
            }
        }

        const authConfigs: ICommandProfileAuthConfig[] = [];
        ImperativeConfig.instance.loadedConfig.profiles?.forEach((profCfg) => {
            if ((profCfg.type === profileType || profCfg.type === "base") && profCfg.authConfig != null) {
                authConfigs.push(...profCfg.authConfig);
            }
        });

        for (const authConfig of authConfigs) {
            const authHandler: ICommandHandlerRequire = require(authConfig.handler);
            const authHandlerClass = new authHandler.default();

            if (authHandlerClass instanceof AbstractAuthHandler) {
                const { promptParams } = authHandlerClass.getAuthHandlerApi();
                if (profile.tokenType === promptParams.defaultTokenType || profile.tokenType.startsWith(TOKEN_TYPE_APIML)) {
                    return authHandlerClass;  // Auth service must have matching token type
                }
            }
        }
    }

    /**
     * Stores session config properties into a team config profile.
     * @param params CLI handler parameters object
     * @param sessCfg Session config containing properties to store
     * @param propsToStore Names of properties that should be stored
     */
    public static async storeSessCfgProps(params: IHandlerParameters, sessCfg: { [key: string]: any }, propsToStore: string[]): Promise<void> {
        return this._storeSessCfgProps({ params, sessCfg, propsToStore });
    }

    /**
     * Stores session config properties into a team config profile.
     * @param opts Set of options required to store session config properties
     */
    public static async _storeSessCfgProps(opts: IConfigAutoStoreStoreSessCfgPropsOpts): Promise<void> {
        const config = opts.config || ImperativeConfig.instance.config;
        // TODO Which autoStore value should take priority if it conflicts between layers
        if (opts.propsToStore.length == 0 || !config?.exists || !config.properties.autoStore) {
            Logger.getAppLogger().info("Skipping update of profile properties. Check that config file exists and autoStore is true.");
            return;
        }

        let profileProps = opts.propsToStore.map(propName => propName === "hostname" ? "host" : propName);
        const profileData = this._findActiveProfile({ ...opts, profileProps });
        if (profileData == null && opts.profileName == null && opts.profileType == null) {
            return;
        }
        const [profileType, profileName] = profileData ?? [opts.profileType, opts.profileName];
        const profilePath = config.api.profiles.getProfilePathFromName(profileName);

        // Replace user and password with tokenValue if tokenType is defined in config
        if (profileProps.includes("user") && profileProps.includes("password") && await this._fetchTokenForSessCfg({ ...opts, profilePath })) {
            profileProps = profileProps.filter(propName => propName !== "user" && propName !== "password");
            profileProps.push("tokenValue");
        }

        const beforeLayer = config.api.layers.get();

        const profileObj = config.api.profiles.get(profileName, false);
        const profileSchema = ImperativeConfig.instance.loadedConfig.profiles?.find(p => p.type === profileType)?.schema;
        const profileSecureProps = config.api.secure.securePropsForProfile(profileName);

        const baseProfileName = ConfigUtils.getActiveProfileName("base", opts.params?.arguments, opts.defaultBaseProfileName);
        const baseProfileObj = config.api.profiles.get(baseProfileName, false);
        const baseProfileSchema = ImperativeConfig.instance.loadedConfig.baseProfile.schema;
        const baseProfileSecureProps = config.api.secure.securePropsForProfile(baseProfileName);

        for (const propName of profileProps) {
            let propProfilePath = profilePath;
            let isSecureProp = profileSchema?.properties[propName]?.secure || profileSecureProps.includes(propName);
            /* If any of the following is true, then property should be stored in base profile:
                (1) Service profile does not exist, but base profile does
                (2) Property is missing from service profile properties/secure objects, but present in base profile
                (3) Property is tokenValue and tokenType is missing from service profile, but present in base profile
                (4) Given profile is just a base profile :yum:
            */
            if ((!config.api.profiles.exists(profileName) && config.api.profiles.exists(baseProfileName)) ||
                (profileObj[propName] == null && !profileSecureProps.includes(propName) &&
                    (baseProfileObj[propName] != null || baseProfileSecureProps.includes(propName))) ||
                (propName === "tokenValue" && profileObj.tokenType == null && baseProfileObj.tokenType != null ||
                profileType === "base")
            ) {
                propProfilePath = config.api.profiles.getProfilePathFromName(baseProfileName);
                isSecureProp = baseProfileSchema.properties[propName].secure || baseProfileSecureProps.includes(propName);
            }

            // If secure array at higher level includes this property, then property should be stored at higher level
            if (isSecureProp) {
                const secureProfile = config.api.secure.secureInfoForProp(`${propProfilePath}.properties.${propName}`, true);
                let secureProfilePath: string;
                if (secureProfile != null) secureProfilePath = secureProfile.path;
                if (secureProfilePath != null && secureProfilePath.split(".").length < propProfilePath.split(".").length) {
                    propProfilePath = secureProfilePath.slice(0, secureProfilePath.lastIndexOf("."));
                }
            }

            const foundLayer = config.api.layers.find(config.api.profiles.getProfileNameFromPath(propProfilePath));
            if (foundLayer != null) config.api.layers.activate(foundLayer.user, foundLayer.global);

            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            config.set(`${propProfilePath}.properties.${propName}`, opts.sessCfg[sessCfgPropName], {
                secure: opts.setSecure ?? isSecureProp
            });
        }

        await config.save();
        const storedMsg = `Stored properties in ${config.layerActive().path}: ${profileProps.join(", ")}`;
        if (opts.params) {
            opts.params.response.console.log(storedMsg);
        } else {
            Logger.getAppLogger().info(storedMsg);
        }
        // Restore original active layer
        config.api.layers.activate(beforeLayer.user, beforeLayer.global);
    }

    /**
     * Retrieves token value that will be auto-stored into session config.
     * @param params CLI handler parameters object
     * @param sessCfg Session config with credentials for basic or cert auth
     * @param profilePath JSON path of profile containing tokenType
     * @returns True if auth handler was found and token was fetched
     */
    private static async fetchTokenForSessCfg(params: IHandlerParameters, sessCfg: { [key: string]: any }, profilePath: string): Promise<boolean> {
        return this._fetchTokenForSessCfg({ params, sessCfg, profilePath });
    }

    /**
     * Helper function that retrieves token value that will be auto-stored into session config
     * @param opts Set of options required to fetch the token value to be auto-stored
     * @returns True if auth handler was found and token was fetched
     */
    private static async _fetchTokenForSessCfg(opts: IConfigAutoStoreStoreSessCfgPropsOpts): Promise<boolean> {
        const authHandlerClass = this._findAuthHandlerForProfile(opts);

        if (authHandlerClass == null) {
            return false;
        }

        const api = authHandlerClass.getAuthHandlerApi();
        opts.sessCfg.type = AUTH_TYPE_TOKEN;
        opts.sessCfg.tokenType = opts.params?.arguments?.tokenType ?? api.promptParams.defaultTokenType;
        const baseSessCfg: ISession = { type: opts.sessCfg.type };

        for (const propName of Object.keys(ImperativeConfig.instance.loadedConfig.baseProfile.schema.properties)) {
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            if (opts.sessCfg[sessCfgPropName] != null) {
                (baseSessCfg as any)[sessCfgPropName] = opts.sessCfg[sessCfgPropName];
            }
        }

        Logger.getAppLogger().info(`Fetching ${opts.sessCfg.tokenType} for ${opts.profilePath}`);
        opts.sessCfg.tokenValue = await api.sessionLogin(new Session(baseSessCfg));
        opts.sessCfg.user = opts.sessCfg.password = undefined;
        return true;
    }
}
