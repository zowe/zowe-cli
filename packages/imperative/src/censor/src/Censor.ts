/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { Arguments } from "yargs";
import { ICommandProfileProperty } from "../../cmd/src/doc/profiles/definition/ICommandProfileProperty";
import { CliUtils } from "../../utilities/src/CliUtils";
import { ICensorOptions } from "./doc/ICensorOptions";
import { Config } from "../../config/src/Config";
import { ImperativeConfig } from "../../utilities/src/ImperativeConfig";
import { EnvironmentalVariableSettings } from "../../imperative/src/env/EnvironmentalVariableSettings";
import { ICommandProfileTypeConfiguration } from "../../cmd/src/doc/profiles/definition/ICommandProfileTypeConfiguration";
import { IProfileSchema} from "../../profiles/src/doc/definition/IProfileSchema";
import { IProfileTypeConfiguration } from "../../profiles/src/doc/config/IProfileTypeConfiguration";

export class Censor {

    /*********************************************************************
    * Basic censorship items - list definitions & initialiazations, etc. *
    **********************************************************************/

    /*
    * NOTE(Kelosky): Ideally we might have a consolidated list for secure fields, but for now we'll just
    * make sure they're collocated within the same class.
    *
    * NOTE(Harn): This list should be kept in sync with the base profile secure definitions and MUST be in camel case.
    */
    private static readonly MAIN_CENSORED_OPTIONS = ["auth", "authentication", "basicAuth", "certFilePassphrase", "credentials",
        "pw", "pass", "password", "passphrase", "tv", "tokenValue"];

    private static readonly MAIN_SECURE_PROMPT_OPTIONS = ["keyPassphrase", "password", "passphrase", "tokenValue", "user"];

    // The censor response.
    public static readonly CENSOR_RESPONSE = "****";

    // A set of default censored options.
    public static get DEFAULT_CENSORED_OPTIONS(): string[] {
        const censoredList = new Set<string>();
        for (const option of this.MAIN_CENSORED_OPTIONS) {
            censoredList.add(option);
            censoredList.add(CliUtils.getOptionFormat(option).kebabCase);
        }
        return Array.from(censoredList);
    }

    // Return a customized list of secure prompt options
    public static get SECURE_PROMPT_OPTIONS(): string[] {
        const censoredList = new Set<string>();
        for (const option of this.MAIN_SECURE_PROMPT_OPTIONS) {
            censoredList.add(option);
            censoredList.add(CliUtils.getOptionFormat(option).kebabCase);
        }
        return Array.from(censoredList);
    }

    // Set a censored options list that can be set and retrieved for each command.
    private static mCensoredOptions: Set<string> = new Set(this.DEFAULT_CENSORED_OPTIONS);

    // Return a customized list of censored options (or just the defaults if not set).
    public static get CENSORED_OPTIONS(): string[] {
        return Array.from(this.mCensoredOptions);
    }

    //Singleton caches of the Config, Command Definition and Command Arguments
    private static mConfig: Config = null;

    /**
     * Singleton implementation of an internal reference to the schema
     */
    private static mSchema: ICommandProfileTypeConfiguration[] = null;

    /**
     * Helper method to get an internal reference to the loaded profiles
     */
    public static get profileSchemas(): ICommandProfileTypeConfiguration[] {
        if (this.mSchema == null) this.mSchema = ImperativeConfig.instance.loadedConfig?.profiles ?? [];
        return this.mSchema;
    }

    /**
     * Helper method to set an internal reference to loaded profiles
     * @param _schemas - The schmas to pass in to set to the logger
     */
    public static setProfileSchemas(_schemas: IProfileTypeConfiguration[] | Map<string, IProfileSchema>) {
        if (this.mSchema == null) {
            this.mSchema = [];
        }
        if (_schemas instanceof Map) {
            _schemas.forEach((v: IProfileSchema) => {
                this.mSchema.push({ type: v.type, schema: v });
            });
        } else if (Array.isArray(_schemas)) {
            _schemas.forEach((v: IProfileTypeConfiguration) => {
                this.mSchema.push({ type: v.type, schema: v.schema });
            });
        }
    }

    /****************************************************
     * Helper functions for more advanced functionality *
     ****************************************************/

    /**
     * Helper function to handle profile schemas when setting the censored options
     * @param {IProfileTypeConfiguration | ICommandProfileTypeConfiguration} profileType - the profile type configuration to iterate over
     */
    private static handleSchema(profileType: IProfileTypeConfiguration | ICommandProfileTypeConfiguration): void {
        const secureOptions: Set<string> = new Set();

        /* eslint-disable-next-line no-unused-vars */
        for (const [key, cmdProp] of Object.entries(profileType.schema.properties)) {
            const prop = cmdProp as ICommandProfileProperty;
            // Add censored options from the schema if the option is secure
            if (prop.secure) {
                // Handle the case of a single option definition
                if (prop.optionDefinition) {
                    secureOptions.add(prop.optionDefinition.name);
                    for (const alias of prop.optionDefinition.aliases || []) {
                        // Remember to add the alias
                        secureOptions.add(alias);
                    }
                } else if (prop.optionDefinitions) {
                    // Handle the case of multiple option definitions
                    prop.optionDefinitions.forEach(opDef => {
                        secureOptions.add(opDef.name);
                        for (const alias of opDef.aliases || []) {
                            // Remember to add the alias
                            secureOptions.add(alias);
                        }
                    });
                } else {
                    secureOptions.add(key);
                }
            }
        }

        secureOptions.forEach(prop => this.addCensoredOption(prop));
    }

    /**
     * Add a censored option, including it's camelCase and kebabCase versions
     * @param {string} option - The option to censor
     */
    private static addCensoredOption(option: string) {
        // This option is required, but we do not want to ever allow null or undefined itself into the censored options
        if (option != null) {
            this.mCensoredOptions.add(option);
            this.mCensoredOptions.add(CliUtils.getOptionFormat(option).camelCase);
            this.mCensoredOptions.add(CliUtils.getOptionFormat(option).kebabCase);
        }
    }

    /**
     * Specifies whether a given property path (e.g. "profiles.lpar1.properties.host") is a special value or not.
     * Special value: Refers to any value defined as secure in the schema definition.
     *                These values should be already masked by the application (and/or plugin) developer.
     * @param prop Property path to determine if it is a special value
     * @returns True - if the given property is to be treated as a special value; False - otherwise
     */
    public static isSpecialValue(prop: string): boolean {
        let specialValues = this.SECURE_PROMPT_OPTIONS;
        const getPropertyNames = (prop: ICommandProfileProperty): string[] => {
            const ret: string[] = [];
            ret.push(prop.optionDefinition?.name);
            prop.optionDefinitions?.map(opDef => ret.push(opDef.name));
            return ret;
        };

        for (const profile of this.profileSchemas) {
            for (const prop of Object.values(profile.schema.properties)) {
                if (prop.secure) specialValues = lodash.union(specialValues, getPropertyNames(prop));
            }
        }

        return specialValues.some((v) => prop.endsWith(`.${v}`));
    }

    /**
     * Identifies if a property is a secure property
     * @param {string} prop - The property to check
     * @returns {boolean} - True if the property is secure; False otherwise
     */
    public static isSecureValue(prop: string) {
        return this.CENSORED_OPTIONS.includes(prop);
    }

    /****************************************************************************************
     * Bread and butter functions, setting up the class and performing censorship of values *
     ****************************************************************************************/

    /**
     * Generate and set the list of censored options.
     * Attempt to source the censored options from the schema, config, and/or command being executed.
     * @param {ICensorOptions} censorOpts - The objects to use to gather options that should be censored
     */
    public static setCensoredOptions(censorOpts?: ICensorOptions) {
        this.mCensoredOptions = new Set(this.DEFAULT_CENSORED_OPTIONS);

        if (censorOpts) {
            // Save off the config object
            this.mConfig = censorOpts.config;

            // If we have a ProfileTypeConfiguration (i.e. ImperativeConfig.instance.loadedConfig.profiles)
            if (censorOpts.profiles) { this.setProfileSchemas(censorOpts.profiles); }

            for (const profileType of this.profileSchemas ?? []) {
                // If we know the command we are running, and we know the profile types that the command uses
                // we should only use those profiles to determine what should be censored. If we do not, we should
                // add everything
                if (censorOpts.commandDefinition == null ||
                    censorOpts.commandDefinition.profile?.optional?.includes(profileType.type) ||
                    censorOpts.commandDefinition.profile?.required?.includes(profileType.type)) {
                    this.handleSchema(profileType);
                }
            }

            // Include any secure options from the config
            if (censorOpts.config) {
                const secureOptions: Set<string> = new Set();

                // Try to use the command and inputs to find the profiles being loaded
                if (censorOpts.commandDefinition && censorOpts.commandArguments) {
                    const profiles = [];
                    for (const prof of censorOpts.commandDefinition.profile?.required || []) {
                        profiles.push(prof);
                    }
                    for (const prof of censorOpts.commandDefinition.profile?.optional || []) {
                        profiles.push(prof);
                    }

                    for (const prof of profiles) {
                        // If the profile exists, append all of the secure props to the censored list
                        let profName = censorOpts.commandArguments?.[`${prof}-profile`];
                        if (!profName) { profName = this.mConfig.mProperties.defaults[`${prof}`]; }
                        if (profName && censorOpts.config.api.profiles.get(profName)) {
                            censorOpts.config.api.secure.securePropsForProfile(profName).forEach(prop => secureOptions.add(prop));
                        }
                    }
                } else {
                    // We only have a configuration file, assume every property that is secured should be censored
                    censorOpts.config.api.secure.findSecure(censorOpts.config.mProperties.profiles, "profiles").forEach(
                        prop => secureOptions.add(prop.split(".").pop())
                    );
                }
                secureOptions.forEach(prop => this.addCensoredOption(prop));
            }
        } else if (this.profileSchemas) {
            for (const profileType of this.profileSchemas) {
                this.handleSchema(profileType);
            }
        }
    }

    /**
     * Copy and censor any sensitive CLI arguments before logging/printing
     * @param {string[]} args - The args list to censor
     * @returns {string[]}
     */
    public static censorCLIArgs(args: string[]): string[] {
        const newArgs: string[] = JSON.parse(JSON.stringify(args));
        const censoredValues = this.CENSORED_OPTIONS.map(CliUtils.getDashFormOfOption);
        for (const value of censoredValues) {
            if (args.indexOf(value) >= 0) {
                const valueIndex = args.indexOf(value);
                if (valueIndex < args.length - 1) {
                    newArgs[valueIndex + 1] = this.CENSOR_RESPONSE; // censor the argument after the option name
                }
            }
        }
        return newArgs;
    }

    /**
     * Copy and censor any sensitive CLI arguments before logging/printing
     * @param {string} data - the data to censor
     * @returns {string} - the censored data
     */
    public static censorRawData(data: string, category: string = ""): string {
        const config = this.mConfig ?? ImperativeConfig.instance?.config;

        // Return the data if not using config
        if (!config?.exists) return data;

        // Return the data if we are printing to the console and masking is disabled
        if (ImperativeConfig.instance?.envVariablePrefix) {
            const envMaskOutput = EnvironmentalVariableSettings.read(ImperativeConfig.instance.envVariablePrefix).maskOutput.value;
            const envShowSecureArgs = EnvironmentalVariableSettings.read(ImperativeConfig.instance.envVariablePrefix).showSecureArgs.value;
            // Hardcoding "console" instead of using Logger.DEFAULT_CONSOLE_NAME because of circular dependencies
            if ((category === "console" || category === "json") &&
                (envMaskOutput.toUpperCase() === "FALSE" || envShowSecureArgs.toUpperCase() === "TRUE")) {
                return data;
            }
        }

        let newData = data;

        const secureFields = config.api.secure.findSecure(config.mProperties.profiles, "profiles");
        for (const prop of secureFields) {
            const sec = lodash.get(config.mProperties, prop);
            if (sec && typeof sec !== "object" && !this.isSpecialValue(prop) && this.isSecureValue(prop.split(".").pop())) {
                newData = newData.replace(new RegExp(sec, "gi"), this.CENSOR_RESPONSE);
            }
        }
        return newData;
    }

    /**
     * Copy and censor a yargs argument object before logging
     * @param {yargs.Arguments} args - the args to censor
     * @returns {yargs.Arguments} - a censored copy of the arguments
     */
    public static censorYargsArguments(args: Arguments): Arguments {
        const newArgs: Arguments = JSON.parse(JSON.stringify(args));

        for (const optionName of Object.keys(newArgs)) {
            if (this.CENSORED_OPTIONS.indexOf(optionName) >= 0) {
                const valueToCensor = newArgs[optionName];
                newArgs[optionName] = this.CENSOR_RESPONSE;
                for (const checkAliasKey of Object.keys(newArgs)) {
                    if (newArgs[checkAliasKey] === valueToCensor) {
                        newArgs[checkAliasKey] = this.CENSOR_RESPONSE;
                    }
                }
            }
        }
        return newArgs;
    }
}