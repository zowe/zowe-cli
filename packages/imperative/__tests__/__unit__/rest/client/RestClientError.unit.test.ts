/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { RestClientError } from "../../../../src/rest/client/RestClientError";
import { ImperativeError } from "../../../../src/error";

describe("RestClientError tests", () => {
    it("should be an instance of ImperativeError", async () => {
        const err = new RestClientError({
            msg: "This is a rest client error",
            additionalDetails: "This is an error message",
            source: "client"
        });
        try {
            throw err;
        } catch (restErr) {
            expect(restErr).toBeInstanceOf(ImperativeError);
        }
    });

    it("should have connection details available", async () => {
        const err = new RestClientError({
            msg: "This is a rest client error",
            additionalDetails: "This is an error message",
            source: "client",
            host: "blah",
            port: 1234,
            resource: "/the/resource",
            request: "GET",
            basePath: "base/path",
            headers: ["header"],
            payload: "the payload",
            httpStatus: 200
        });
        try {
            throw err;
        } catch (restErr) {
            expect(restErr.details).toMatchSnapshot();
        }
    });
});
