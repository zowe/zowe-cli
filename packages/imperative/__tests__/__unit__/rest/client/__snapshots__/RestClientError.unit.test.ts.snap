// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestClientError tests should have connection details available 1`] = `
Object {
  "additionalDetails": "This is an error message",
  "basePath": "base/path",
  "headers": Array [
    "header",
  ],
  "host": "blah",
  "httpStatus": 200,
  "msg": "This is a rest client error",
  "payload": "the payload",
  "port": 1234,
  "request": "GET",
  "resource": "/the/resource",
  "source": "client",
}
`;
